pkg/util/bitarray/bitarray.go

```diff
@@ -538,3 +538,37 @@ func Next(d BitArray) BitArray {
 	copy(res.words, d.words)
 	return res
 }
+
+// GetBitAtIndex extract bit at given index in the BitArray.
+func (d BitArray) GetBitAtIndex(index int) (int, error) {
+	// Checking whether index asked is inside BitArray.
+	if index < 0 || uint(index) >= d.BitLen() {
+		err := fmt.Errorf("GetBitAtIndex: bit index %d out of valid range (0..%d)", index, d.BitLen()-1)
+		return 0, pgerror.WithCandidateCode(err, pgcode.ArraySubscript)
+	}
+	// To extract bit at the given index, we have to determine the
+	// position within words array, i.e. index/numBitsPerWord after
+	// that checked the bit at residual index.
+	if d.words[index/numBitsPerWord]&(word(1)<<(numBitsPerWord-1-uint(index)%numBitsPerWord)) != 0 {
+		return 1, nil
+	}
+	return 0, nil
+}
+
+// SetBitAtIndex returns the BitArray with an updated bit at a given index.
+func (d BitArray) SetBitAtIndex(index, toSet int) (BitArray, error) {
```

Comment:
hi @rohany,
I am had tried SetBitAtIndex method without a deep copy, it caused a problem in UNION operation,

as when I did ```SELECT set_bit(B'1101010', 0, 0) UNION SELECT set_bit(B'1101010', 2, 1)```
for this test case it only generated output ```0111010```
single row only

Reply:
we should deep copy, otherwise something like 

```
otan=# create table ba_test(a varbit);
CREATE TABLE
otan=# insert into ba_test values ('00');
INSERT 0 1
otan=# select set_bit(a, 1 ,1), get_bit(a, 1), set_bit(a, 1, 0), get_bit(a, 1) from ba_test;
 set_bit | get_bit | set_bit | get_bit
---------+---------+---------+---------
 01      |       0 | 00      |       0
(1 row)
```
would get inconsistent results.